src/Microsoft.SqlTools.ServiceLayer/Extensibility/ExtensionServiceProvider.cs

```diff
@@ -0,0 +1,184 @@
+        }
+
+        protected override IEnumerable<T> GetServicesImpl<T>()
+        {
+            EnsureExtensionStoreRegistered<T>();
+            return base.GetServicesImpl<T>();
+        }
+
+        private void EnsureExtensionStoreRegistered<T>()
+        {
+            if (!services.ContainsKey(typeof(T)))
+            {
+                ExtensionStore store = new ExtensionStore(typeof(T), config);
+                base.Register<T>(() => store.GetExports<T>());
+            }
+        }
+    }
+    
+    /// <summary>
+    /// A store for MEF exports of a specific type. Provides basic wrapper functionality around MEF to standarize how
+    /// we lookup types and return to callers.
+    /// </summary>
+    public class ExtensionStore
+    {
+        CompositionHost host;
+        private IList exports;
+        private Type contractType;
+        
+        /// <summary>
+        /// Initializes the store with a type to lookup exports of, and a function that configures the
+        /// lookup parameters.
+        /// </summary>
+        /// <param name="contractType">Type to use as a base for all extensions being looked up</param>
+        /// <param name="configure">Function that returns the configuration to be used</param>
+        public ExtensionStore(Type contractType, Func<ConventionBuilder, ContainerConfiguration> configure)
+        {
+            Validate.IsNotNull(nameof(contractType), contractType);
+            Validate.IsNotNull(nameof(configure), configure);
+            this.contractType = contractType;
+            ConventionBuilder builder = GetExportBuilder();
+            ContainerConfiguration config = configure(builder);
+            host = config.CreateContainer();
+        }
+
+        /// <summary>
+        /// Loads extensions from the current assembly
+        /// </summary>
+        /// <returns>ExtensionStore</returns>
+        public static ExtensionStore CreateDefaultLoader<T>()
+        {
+            return CreateAssemblyStore<T>(typeof(ExtensionStore).GetTypeInfo().Assembly);
+        }
+        
+        public static ExtensionStore CreateAssemblyStore<T>(Assembly assembly)
+        {
+            Validate.IsNotNull(nameof(assembly), assembly);
+            return new ExtensionStore(typeof(T), (conventions) =>
+                new ContainerConfiguration().WithAssembly(assembly, conventions));
+        }
+
+        public static ExtensionStore CreateStoreForCurrentDirectory<T>()
+        {
+            string assemblyPath = typeof(ExtensionStore).GetTypeInfo().Assembly.Location;
+            string directory = Path.GetDirectoryName(assemblyPath);
+            return new ExtensionStore(typeof(T), (conventions) => 
+                new ContainerConfiguration().WithAssembliesInPath(directory, conventions));
+        }
+
+        public IEnumerable<T> GetExports<T>()
+        {
+            if (exports == null)
+            {
+                exports = host.GetExports(contractType).ToList();
```

Comment:
No - it's an enumerable and returns empty enum rather than null.

---
In reply to: [101187416](https://github.com/Microsoft/sqltoolsservice/pull/229#discussion_r101187416) [](ancestors = 101187416)